tools/undelete-blocks/main.go

```diff
@@ -0,0 +1,471 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/grafana/dskit/flagext"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/storage/tsdb/block"
+	"github.com/grafana/mimir/pkg/util/objtools"
+)
+
+type config struct {
+	bucketConfig    objtools.BucketConfig
+	includeTenants  flagext.StringSliceCSV
+	excludeTenants  flagext.StringSliceCSV
+	inputFile       string
+	inputFileFormat string
+	dryRun          bool
+}
+
+func (c *config) registerFlags(f *flag.FlagSet) {
+	c.bucketConfig.RegisterFlags(f)
+	f.Var(&c.includeTenants, "include-tenants", "If not empty, only blocks for these tenants are considered.")
+	f.Var(&c.excludeTenants, "exclude-tenants", "Blocks for these tenants are not considered.")
```

Comment:
I would remove tenant filtering functionality. This tool should restore all supplied blocks.

Reply:
The tenant filtering was very intentionally added and has a purpose. For the listing option it gives targeting. For the `lines` and `json` options it allows a large input to be reused without externally modifying it.